cli/module_loader.rs

```diff
@@ -166,61 +157,40 @@ impl ModuleLoadPreparer {
       )
       .await?;
 
-    // If there is a lockfile, validate the integrity of all the modules.
+    graph_valid_with_cli_options(graph, &roots, &self.options)?;
+
+    // If there is a lockfile...
     if let Some(lockfile) = &self.lockfile {
-      graph_lock_or_exit(graph, &mut lockfile.lock());
+      let mut lockfile = lockfile.lock();
+      // validate the integrity of all the modules
+      graph_lock_or_exit(graph, &mut lockfile);
+      // update it with anything new
+      lockfile.write()?;
     }
 
-    graph_valid_with_cli_options(graph, &roots, &self.options)?;
     // save the graph and get a reference to the new graph
     let graph = graph_update_permit.commit();
 
-    if graph.has_node_specifier
-      && self.options.type_check_mode() != TypeCheckMode::None
-    {
-      self
-        .npm_resolver
-        .inject_synthetic_types_node_package()
-        .await?;
-    }
-
     drop(_pb_clear_guard);
 
     // type check if necessary
     if self.options.type_check_mode() != TypeCheckMode::None
       && !self.graph_container.is_type_checked(&roots, lib)
     {
-      // todo(dsherret): consolidate this with what's done in graph_util
-      log::debug!("Type checking.");
-      let maybe_config_specifier = self.options.maybe_config_file_specifier();
       let graph = Arc::new(graph.segment(&roots));
-      let options = check::CheckOptions {
-        type_check_mode: self.options.type_check_mode(),
-        debug: self.options.log_level() == Some(log::Level::Debug),
-        maybe_config_specifier,
-        ts_config: self
-          .options
-          .resolve_ts_config_for_emit(TsConfigType::Check { lib })?
-          .ts_config,
-        log_checks: true,
-        reload: self.options.reload_flag()
-          && !roots.iter().all(|r| reload_exclusions.contains(r)),
-      };
-      let check_cache =
-        TypeCheckCache::new(self.caches.type_checking_cache_db(&self.deno_dir));
-      let check_result =
-        check::check(graph, &check_cache, self.npm_resolver.clone(), options)?;
+      self
+        .type_checker
+        .check(
+          graph,
+          check::CheckOptions {
+            lib,
+            log_ignored_options: false,
+            reload: self.options.reload_flag()
+              && !roots.iter().all(|r| reload_exclusions.contains(r)),
+          },
+        )
+        .await?;
       self.graph_container.set_type_checked(&roots, lib);
-      if !check_result.diagnostics.is_empty() {
-        return Err(anyhow!(check_result.diagnostics));
-      }
-      log::debug!("{}", check_result.stats);
-    }
-
-    // any updates to the lockfile should be updated now
```

Comment:
We can do this before type checking because it reduces the chance of `@types/node` ending up in the lockfile.